src/porepy/models/compositional_flow.py

```diff
@@ -0,0 +1,2085 @@
+    re-discretization if flagged True. By default, they are assumed to be False.
+
+    The flux discretizations are performed in :meth:`before_nonlinear_iteration`
+    **before** the super-call. Note that this is critical since upwinding (which is
+    expected to be part of the super-call) must be re-discretized **after** the fluxes
+    are re-discretized.
+
+    Notes:
+
+        1. Re-discretizing the MPFA is expensive and will slow down the simulation
+           noticeably.
+        2. When mixing in this class, it must be above other, fully functional solution
+           strategies in order for the super-call to work as intended.
+
+    """
+
+    fourier_flux_discretization: Callable[[list[pp.Grid]], pp.ad.MpfaAd]
+    """See :class:`~porepy.models.constitutive_laws.FouriersLaw`."""
+    darcy_flux_discretization: Callable[[list[pp.Grid]], pp.ad.MpfaAd]
+    """See :class:`~porepy.models.constitutive_laws.DarcysLaw`."""
+
+    def __init__(self, params: Optional[dict] = None) -> None:
+
+        assert isinstance(
+            self, pp.SolutionStrategy
+        ), "This is a mixin. Require SolutionStrategy as base."
+        super().__init__(params)  # type:ignore[safe-super]
+
+        self._nonlinear_flux_discretizations: list[pp.ad._ad_utils.MergedOperator] = []
+        """Separate container for fluxes which need to be re-discretized. The separation
+        is necessary due to the re-discretization being performed at different stages
+        of the algorithm."""
+
+    def add_nonlinear_flux_discretization(
+        self, discretization: pp.ad._ad_utils.MergedOperator
+    ) -> None:
+        """Add an entry to the list of non-linear flux discretizations.
+
+        Parameters:
+            discretization: The nonlinear discretization to be added.
+
+        """
+        if discretization not in self._nonlinear_flux_discretizations:
+            self._nonlinear_flux_discretizations.append(discretization)
+
+    def set_nonlinear_discretizations(self) -> None:
+        """After the super-call, this method adds the
+        :meth:`fourier_flux_discretization` and the :meth:`darcy_flux_discretization`
+        to the update framework using :meth:`add_nonlinear_flux_discretization`."""
+        assert isinstance(
+            self, pp.SolutionStrategy
+        ), "This is a mixin. Require SolutionStrategy as base."
+        super().set_nonlinear_discretizations()  # type:ignore[safe-super]
+
+        subdomains = self.mdg.subdomains()
+
+        if self.params.get("rediscretize_fourier_flux", False):
+            self.add_nonlinear_flux_discretization(
+                self.fourier_flux_discretization(subdomains).flux()
+            )
+        if self.params.get("rediscretize_darcy_flux", False):
+            self.add_nonlinear_flux_discretization(
+                self.darcy_flux_discretization(subdomains).flux()
+            )
+
+    def rediscretize_fluxes(self) -> None:
```

Comment:
Is there any reason not to use the method rediscretize() in solution_strategy?

Reply:
Resolve? See your next comment

Reply:
Put on hold, pending resolution of the next question.

Reply:
Resolving this one, discussion moved to the next question.